packages/plugin-ext/src/common/connection.ts

```diff
@@ -0,0 +1,101 @@
+/********************************************************************************
+ * Copyright (C) 2018 Red Hat, Inc. and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+import { AbstractMessageReader, DataCallback } from 'vscode-jsonrpc/lib/messageReader';
+import { AbstractMessageWriter } from 'vscode-jsonrpc/lib/messageWriter';
+import { ConnectionMain, ConnectionExt } from '../api/plugin-api';
+import { Disposable } from './disposable-util';
+
+export class PluginConnection implements Disposable {
+    reader: PluginMessageReader;
+    writer: PluginMessageWriter;
+    _dispose: () => void;
+
+    constructor(protected readonly pluginMessageReader: PluginMessageReader,
+        protected readonly pluginMessageWriter: PluginMessageWriter,
+        _dispose: () => void) {
+        this.reader = pluginMessageReader;
+        this.writer = pluginMessageWriter;
+        this._dispose = _dispose;
+    }
+
+    dispose(): void {
+        this._dispose();
+    }
+}
+
+export class PluginMessageWriter extends AbstractMessageWriter {
```

Comment:
could we move out these two inner classes(PluginMessageWriter PluginMessageReader ) and  to separate files  as they're exported?

Reply:
divided